### ISSUE TITLE
1-to-1 Inputs and Outputs with matching timestamps are not considered up-to-date

### ISSUE THREAD

**Issue by apodtele** [dotnet/msbuild#7021]
A Target simply copies multiple SourceFiles into DestinationFiles that are also defined as Inputs and Outputs. A Copy task preserves timestamps (undocumented feature). Notice that 1-to-1 pairs will have matching timestamps, which may be different in different pairs. This is not good enough to skip the Target next time despite what [the docs](https://docs.microsoft.com/en-us/visualstudio/msbuild/incremental-builds?view=vs-2022) say. 

Matching timestamps equal in pairs but not throughout are not considered up-to-date. Everything works if timestamps are equal everywhere.

**Comment by BenVillalobos** [comment]
Team Triage: Could you provide a minimal project that reproduces this behavior?

**Comment by apodtele** [comment]
I could not attach it so see below. `TouchSrcTogether` creates the same timestamps so that `CopySrcToDst` copies them once and skips next time. `TouchSrcSeparately` creates files with different timestamps and `CopySrcToDst` never skips.

MSBuild -t:TouchSrcSeparately Test.proj
  Touching "Src1".
  Touching "Src2".
MSBuild -t:CopySrcToDst Test.proj
  Copying file from "C:\Test\Src1" to "C:\Test\Dst1".
  Copying file from "C:\Test\Src2" to "C:\Test\Dst2".
MSBuild -t:CopySrcToDst Test.proj
  Copying file from "C:\Test\Src1" to "C:\Test\Dst1".
  Copying file from "C:\Test\Src2" to "C:\Test\Dst2".

vs.

MSBuild -t:TouchSrcTogether Test.proj
  Touching "Src1".
  Touching "Src2".
MSBuild -t:CopySrcToDst Test.proj
  Copying file from "C:\Test\Src1" to "C:\Test\Dst1".
  Copying file from "C:\Test\Src2" to "C:\Test\Dst2".
MSBuild -t:CopySrcToDst Test.proj
Skipping target "CopySrcToDst" because all output files are up-to-date with respect to the input files.


```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <Src1 Include="Src1" />
    <Src2 Include="Src2" />
    <Src Include="@(Src1);@(Src2)" />
    <Dst Include="Dst1;Dst2" />
  </ItemGroup>
  <Target Name="TouchSrcSeparately">
    <!-- Important: two touches result in different timestamps -->
    <Touch Files="@(Src1)" AlwaysCreate="true" />
    <Touch Files="@(Src2)" AlwaysCreate="true" />
  </Target>
  <Target Name="TouchSrcTogether">
    <!-- Important: one touch results in matching timestamps -->
    <Touch Files="@(Src)" AlwaysCreate="true" />
  </Target>
  <Target Name="TouchDst">
    <Touch Files="@(Dst)" />
  </Target>
  <Target Name="CopySrcToDst" Inputs="@(Src)" Outputs="@(Dst)">
    <Copy SourceFiles="@(Src)" DestinationFiles="@(Dst)" />
  </Target>
</Project>
```

**Comment by BenVillalobos** [comment]
Team Triage: We think the issue is that the inputs and outputs are two separate items that are technically unrelated, so there's not a 1 to 1 mapping.

> 1-to-1 mappings are typically produced by item transformations

If you use item transforms on the _same item_, you should get the desired result. Something similar to what's in the doc you linked: `"@(Src->'%(Identity).%(Extension)')"`

**Comment by apodtele** [comment]
Absolutely not. Even with transformation that impose 1-to-1 relationship technically, the time stamps in pairs are compared incorrectly. Try the project below.

MSBuild -t:TouchSrcSeparately Test.proj
  Touching "Src1".
  Touching "Src2".
MSBuild -t:CopySrcToDst Test.proj
  Copying file from "Src2" to "CopyOfSrc2".
  Copying file from "Src1" to "CopyOfSrc1".
MSBuild -t:CopySrcToDst Test.proj
  Copying file from "Src2" to "CopyOfSrc2".
  Copying file from "Src1" to "CopyOfSrc1".

```
<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup>
    <Src1 Include="Src1" />
    <Src2 Include="Src2" />
    <Src Include="@(Src1);@(Src2)" />
    <!-- This is certainly 1-to-1 -->
    <Dst Include="@(Src->'CopyOf%(Identity)')" />
  </ItemGroup>
  <Target Name="TouchSrcSeparately">
    <!-- Important: two touches result in different timestamps -->
    <Touch Files="@(Src1)" AlwaysCreate="true" />
    <Touch Files="@(Src2)" AlwaysCreate="true" />
  </Target>
  <Target Name="TouchSrcTogether">
    <!-- Important: two touches result in different timestamps -->
    <Touch Files="@(Src)" AlwaysCreate="true" />
  </Target>
  <Target Name="TouchDst">
    <Touch Files="@(Dst)" />
  </Target>
  <Target Name="CopySrcToDst" Inputs="@(Src)" Outputs="@(Dst)">
    <Copy SourceFiles="@(Src)" DestinationFiles="@(Dst)" />
  </Target>
</Project>
```

**Comment by BenVillalobos** [comment]
@apodtele You still have two unique items. Can you modify your target such that the outputs is `"@(Src->'CopyOf%(Identity)')"`?

**Comment by apodtele** [comment]
@BenVillalobos Yes, indeed, with Outputs transformed, the timestamps are paired and checked properly. So the question becomes what "typically" really means. Perhaps it should say

> 1-to-1 mappings are *only* produced by item transformations in Outputs.

Or, maybe there is a way to declare the items paired?

**Comment by BenVillalobos** [comment]
@apodtele Awesome! One of the first things we realized when looking this over is the docs likely need to be updated. cc @ghogen . 

Also cc @rainersigwald in case the doc fix is more involved.
